integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala

```diff
@@ -0,0 +1,143 @@
+package org.constellation
+
+import java.security.KeyPair
+
+import cats.effect.{ContextShift, IO, Timer}
+import cats.syntax.all._
+import com.typesafe.scalalogging.StrictLogging
+import fs2.{Pipe, Stream}
+import io.circe.syntax._
+import org.constellation.keytool.KeyStoreUtils
+import org.constellation.schema.address.AddressCacheData
+import org.constellation.schema.transaction.{LastTransactionRef, Transaction}
+import org.constellation.wallet.{TransactionExt, Wallet}
+import org.http4s.circe.CirceEntityDecoder._
+import org.http4s.circe.CirceEntityEncoder._
+import org.http4s.client.blaze.BlazeClientBuilder
+import org.http4s.client.dsl.io._
+import org.http4s.dsl.io._
+import org.http4s.{EntityDecoder, Request, Uri}
+import org.mockito.IdiomaticMockito
+import org.scalatest.GivenWhenThen
+import org.scalatest.concurrent.Eventually
+import org.scalatest.freespec.AnyFreeSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.matchers.{MatchResult, Matcher}
+import org.scalatest.time.{Seconds, Span}
+
+import scala.concurrent.ExecutionContext.global
+
+class TransactionsProcessingSpec
+    extends AnyFreeSpec
+    with StrictLogging
+    with GivenWhenThen
+    with Matchers
+    with TestConfig
+    with Eventually
+    with IdiomaticMockito {
+
+  implicit val cs: ContextShift[IO] = IO.contextShift(global)
+  implicit val timer: Timer[IO] = IO.timer(global)
+
+  override implicit val patienceConfig: PatienceConfig =
+    PatienceConfig(
+      timeout = scaled(Span(60, Seconds)),
+      interval = scaled(Span(1, Seconds))
+    )
+
+  private val containTheSameElements = (not be empty).and(Matcher { (left: Seq[_]) =>
+    MatchResult(
+      left.forall(_ == left.head),
+      left + " did not contain the same elements",
+      left + " did contain the same elements"
+    )
+  })
+
+  "sent transactions should change src and dst address balances" in {
```

Comment:
a single test scenario should be a separate app which can be run independently.

Reply:
What do you exactly mean by independently? I don't see why this can't be run independently from the rest of the code.

Reply:
I think we are kind of pulling in two different directions here. One is sanity checking code base which I think we can do with simple integration tests formulated in such a way. The other is creating an infrastructure for running more complex tests, that can be run arbitrarily in different configurations, for different branches, on demand or automatically. I can see scenario written in scala as above as a part of a test, but I think we also need a part that manages and sets the state for the whole test, and then runs such scala app with a scenario or a set of scenarios. I mean I still think we will need to have the ability to ssh to an instance to for example change firewall, or grep logs to check if something happened and if we can proceed with next steps e.g. In the perfect world it would be great to have such an infrastructure that we can run it with some scenarios with GitHub workflow automatically for example for a PR and with others more complex ones on demand with a GitHub action or as a separate AWS environment. Github jobs seem ok at first glance, but also I'm not sure how would we use it to also spin up the whole infrastructure and if it can be done completely automatically. To summarize I think what we have in this PR is a good start, but we need all the other parts to put it together and then verify if this part is still compatible. For sure I think we need some TestManager instance that would conduct the test scenarios and probably that would require us to use some scripting language to do that.

Reply:
So are you talking about having your own Jenkins-like instance, that would run the job periodically?

@TheMMaciek all of the things you mentioned, like logging via SSH into an instance or manipulating firewall _can_ be done with scenarios run with GitHub CI. I can't imagine a scenario that could be run on Jenkins or a "TestManager" deployed to EC2 instance, that couldn't be run on GitHub CI.

Reply:
I think so - and maybe not about running tests periodically but running checks periodically (that's for sure) and tests on demand, or scheduled in some way, or a set of tests one after another. I think the problem is not with a single scenario but with everything combined together. I even thought about a custom CLI/UI tool which will manage tests, make summaries, etc. but it's for future.

tl;dr; you can think about such a scenario:

--- **single scenario but long**:
- Send transactions (**not** via the load balancer) from the same source address to different nodes
- Check if all transactions has been processed (based on the tx amount - this can lasts even few hours)

--- **more complex scenario**:
- Use 10 nodes
- Keep sending transactions (i.e. ~100tx/min) via the load balancer
- Wait for snapshots
- Simulate network issue on 1 node
- Check nodegrid for status
- Stop two nodes one after the other
- Check nodegrid for status
- Reconnect that two nodes
- ...

--- **set of scenarios**:
- Check single tx scenario
- Check multiple txs via load balancer
- Check multiple txs mixed to nodes
- Check blacklisting
- Check joining and dropping
- Check network issues
- Check redownloads
- Periodically check snapshot alignment
- Periodically check majority heights
- ...

Such scenarios/set of scenarios can be long. We should be able to run selected scenarios (not always all of them). Personally I think that it will be really hard to do using only GitHub CI, with no intermediate instance/manager or sth like that.

Reply:
I don't mean periodically. If we checked the branch once there is no need to check it again unless there are changes added I guess. I would for example give a possibility to run the tests, or a selected test scenario on a feature branch so the developer can pre test his changes. If you say that we can do everything with GitHub CI then great! Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?

Reply:
So the longest scenario you mentioned:

>     Send transactions (not via the load balancer) from the same source address to different nodes
>     Check if all transactions has been processed (based on the tx amount - this can lasts even few hours)

is the one that's implemented with this change, but currently on a small volume. It sends 100 transactions to 2 nodes and so far it takes only 1 minute. Even if we would go for 10k transactions over 3 nodes it still wouldn't be very long. I wouldn't go for more in this scenario because it's not a performance test, rather a consensus test. For performance tests, we will be likely using different source addresses, so the throughput will be higher, resulting in shorter runs.

Reply:
but it's not the only scenario which can take longer time and currently it's no time to think about all possible scenarios and calculate how long they can run. The original ticket has a name:
> Long running cluster tests

and that's the business issue/value we want to be solved. Assuming we have tests which can take hours or days, we need a way to run them.

Simple integration tests are something different than long running tests which we should focus on.

Reply:
Ok, so for now, I would suggest that we proceed with GitHub CI and if there will be ever a need to run test that will take a week or a month to complete we will go for [self-hosted runner](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners) on EC2. Is it fine with you?

Reply:
>Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?

@TheMMaciek If the test logic could be expressed with scala, then I think we would just add another scenario to `integration-test/src/test/scala` and it will be picked automatically by sbt runner. If we however need to use another tool, then I would just add another step to the existing workflow, instead of creating a new one.

Reply:
>>>
Workflow run time - Each workflow run is limited to 72 hours. If a workflow run reaches this limit, the workflow run is cancelled.

>>>
Job queue time - Each job for self-hosted runners can be queued for a maximum of 24 hours. If a self-hosted runner does not start executing the job within this limit, the job is terminated and fails to complete.

>>>
API requests - You can execute up to 1000 API requests in an hour across all actions within a repository. If exceeded, additional API calls will fail, which might cause jobs to fail.

Still you would need a manager for this and somehow schedule jobs on CI. Anyway, what's your idea to schedule a scenario set using Github CI only? How will we be able to specify which scenarios we would like to run? How results will be summarised? IMO that's the things we should focus on before even writing a single scenario.

Reply:
> > Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?
> 
> @TheMMaciek If the test logic could be expressed with scala (note that there is a scala SSH client and java AWS SDK), then I think we would just add another scenario to `integration-test/src/test/scala` and it will be picked automatically by sbt runner. If we however need to use another tool, then I would just add another step to the existing workflow, instead of creating a new one.

I told you about scenarios for which we will use something different than scala because scala wouldn't be so efficient. Also, we would like to specify which scenarios should be ran, which no. Writing only one github job will end up with a really long single test with almost no business value for us.

Reply:
>Anyway, what's your idea to schedule a scenario set using Github CI only?

Not sure if I understand the question, but there are [scheduled jobs](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#schedule) on GitHub CI.

>How will we be able to specify which scenarios we would like to run?

By default, I would just run all the scenarios every single time, but if there will be a need to just run some of them, we could categorize them by [tags in ScalaTest](https://www.scalatest.org/user_guide/tagging_your_tests).

>How results will be summarised?

We get the reports from the ScalaTest, or any other testing tool that we would use (all of the testing tools I know, have reporting tool built-in).

Reply:
Resolved, for now we use `ScalaTest` for reports and for running tagged tests.